### DISCUSSION TITLE
State Indicator Plug-In?

### DISCUSSION THREAD

**Discussion by phil-barrett** [terjeio/grblHAL#287] [Ideas]
For a while, I have been thinking about how to do various indicators (think LEDs, for now) with grblHAL.  There is a related discussion about supporting various lighted switches. The discussion about I/O extenders with Pico port gives one possible approach. It is possible to write a plugin that monitors the current machine state and sets the appropriate LED.

I would like to design a board to support this - I have a test extender card that fits on the I2C header of the T41 BOB. .  My thinking is an I2C based board with necessary driver circuitry for 5 or 12V indicator circuits that would fit on the I2C header of the T41 BOB. 

Perhaps someone would like to take a shot at writing the plug-in? It would basically monitor the system state and set outputs on the extender. A simple configuration file (like my_machine.h) could define the states and actions? 

![IMG_20210410_084246](https://user-images.githubusercontent.com/16570958/114275900-230dec00-99d9-11eb-9571-9378b3281d64.jpg)

**Comment by mundsen** [comment]
An alternative?

I2c mini Oled display

https://www.tme.eu/no/en/details/df-dfr0650/other-modules/dfrobot/dfr0650/?brutto=0&gclid=Cj0KCQjwmcWDBhCOARIsALgJ2QcXgnLlV4n5v2i7KTPNe99la4zz653r2KQMG69f40FjuprS1dV9OpEaAt6xEALw_wcB

**Reply by phil-barrett** [reply]
Yes, a display is very desirable.  Though, I think it may be better to have a microcontroller controlling the display and the plug-in sending state information to it. The plug-in would not be responsible for the display, just suppling the data to the display controller.  Ideally, the display would be located on the outside of the control box rather than close to the board. This brings up another issue, though. The I2C bus should not be very long.  But, more significantly, it can act as an EMI antenna - bad in a machine environment.  A general design rule is to keep the I2C bus short.

A display panel is something that I am looking at but is farther down my stack. Also, [Terje has something](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack) that is much farther along and a better starting point.

**Comment by dresco** [comment]
I've also been poking around at a control panel. Early days, but my initial hacky prototype seemed to work well enough (for a mess of wires on the bench). Am using modbus for the comms, as it will be outside of the cabinet. 

Prototype had both inputs (buttons, encoders, pots) and outputs (LEDs, display) sent over modbus registers. Think jogging controls, overrides, DRO & status display etc. Am still planning on using the dedicated inputs for e-stop, cycle start, and feed hold.

My next step is laying out a board for the panel. For an easy low volume prototype, am planning to use an STM32 Nucleo, with PCB mounted tactile switches for the majority of the buttons - which can then sit behind a home printed fascia panel. In my mind, am thinking something like a mini Haas panel, keeping it to A4 size for ease of printing the overlay.

**Reply by phil-barrett** [reply]
This sound really great!  Do you have a plug-in for grblHAL yet?  I have an Async connector on the T41 BOBs.  The Pro version has an extra pin for signaling.  What kind of baud rate do you need?  I've tested to 250K with no problems.  Goal is at least 1Mbaud

**Reply by dresco** [reply]
Hi, yes. Is pretty rough PoC plugin code, but from memory I ended up at 25ms refresh intervals, running the serial at 115k. Am alternating input and output registers with each exchange, so effectively 20Hz update for each. Seemed plenty responsive enough..

My only Teensy/BOB is in the router cabinet, so all of this was prototyped on the bench with bare STM32 boards. Originally just TTL hookups between the boards, but now my cheapo RS485 modules have arrived. They're the automatic flow control ones, so hoping will be a seamless addition.
